### PR TITLE
rgw: eradicate dynamic memory allocation in RGWWriteRequest.

### PR DESCRIPTION
**Pull request by rzarzynski**
It would be really nice to have Boost 1.56 that brought `boost::optional::emplace()`.

CC: @mattbenjamin.

### CODE REVIEW

File: src/rgw/rgw_file.cc [hunk 1]
```diff
@@ -987,9 +995,17 @@ namespace rgw {
       data.swap(orig_data);
 
       /* restart processing with different oid suffix */
-      dispose_processor(processor);
-      processor = select_processor(*static_cast<RGWObjectCtx *>(s->obj_ctx),
-				   &multipart);
+      processor = boost::none;
```
Comment (cbodley) [inline src/rgw/rgw_file.cc]: is this really needed? does `boost::optional` try to call `RGWPutObjProcessor_Atomic::operator=()` otherwise?

**Comment by cbodley** [comment]
@rzarzynski have you looked at `RGWPutObj::execute()`? would the same strategy work there?

changing it now would cause conflicts with https://github.com/ceph/ceph/pull/10932, but we should get that eventually

**Comment by rzarzynski** [comment]
@cbodley: I'm afraid that for `RGWPutObj::execute()` we would need `static_ptr` to handle the polymorphism. The `RGWPutObj` instantiates `RGWPutObjProcessor_Atomic` or `RGWPutObjProcessor_Multipart` depending on the `multipart` variable.

**Comment by yehudasa** [comment]
@cbodley ping

**Comment by dachary** [comment]
jenkins test this please (no logs)

**Comment by cbodley** [comment]
@rzarzynski should be able to use `optional::emplace()` here now?

**Comment by rzarzynski** [comment]
@cbodley: yep, we should be perfectly able to do the switch now. I will send the update soon.

**Comment by rzarzynski** [comment]
@cbodley, @mattbenjamin: fixed the commit and switched to `boost::optional::emplace`.

**Comment by mattbenjamin** [comment]
@rzarzynski I'm fine with using this strategy here, will test

**Comment by rzarzynski** [comment]
@mattbenjamin: thanks! :-)

**Comment by oritwas** [comment]
@rzarzynski please rebase

**Comment by yehudasa** [comment]
@rzarzynski what's the status of this one?

**Comment by amitkumar50** [comment]
https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#title-of-pull-requests-and-title-of-commits

**Comment by cbodley** [comment]
see https://github.com/ceph/ceph/pull/24453

**Pull request closed without merging**